src/pkg/oci/pull.go

```diff
@@ -0,0 +1,241 @@
+	}
+	estimatedBytes += manifest.Config.Size
+
+	dst, err := file.New(destinationDir)
+	if err != nil {
+		return err
+	}
+	defer dst.Close()
+
+	copyOpts := o.CopyOpts
+	copyOpts.Concurrency = concurrency
+	if isPartialPull {
+		paths := []string{}
+		for _, layer := range layersToPull {
+			paths = append(paths, layer.Annotations[ocispec.AnnotationTitle])
+		}
+		copyOpts.FindSuccessors = func(ctx context.Context, fetcher content.Fetcher, desc ocispec.Descriptor) ([]ocispec.Descriptor, error) {
+			nodes, err := content.Successors(ctx, fetcher, desc)
+			if err != nil {
+				return nil, err
+			}
+			var ret []ocispec.Descriptor
+			for _, node := range nodes {
+				if utils.SliceContains(paths, node.Annotations[ocispec.AnnotationTitle]) {
+					ret = append(ret, node)
+				}
+			}
+			return ret, nil
+		}
+	}
+
+	// Create a thread to update a progress bar as we save the package to disk
+	doneSaving := make(chan int)
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go utils.RenderProgressBarForLocalDirWrite(destinationDir, estimatedBytes, &wg, doneSaving, "Pulling Zarf package data")
+	_, err = oras.Copy(o.Context, o.Repository, ref.String(), dst, ref.String(), copyOpts)
+	if err != nil {
+		return err
+	}
+
+	// Send a signal to the progress bar that we're done and wait for it to finish
+	doneSaving <- 1
+	wg.Wait()
+
+	message.Debugf("Pulled %s", ref.String())
+	message.Successf("Pulled %s", ref.String())
+
+	layersToCheck := []string{}
+	if isPartialPull {
+		for _, layer := range layersToPull {
+			if layer.Annotations[ocispec.AnnotationTitle] != "" {
+				layersToCheck = append(layersToCheck, layer.Annotations[ocispec.AnnotationTitle])
+			}
+		}
+	}
+	return utils.ValidatePackageChecksums(destinationDir, layersToCheck)
+}
+
+// PullFileLayer pulls a file layer from the remote repository and saves it to `destinationDir/annotationTitle`.
+func (o *OrasRemote) PullFileLayer(desc ocispec.Descriptor, destinationDir string) error {
+	if desc.MediaType != ZarfLayerMediaTypeBlob {
+		return fmt.Errorf("invalid media type for file layer: %s", desc.MediaType)
+	}
+	b, err := o.FetchLayer(desc)
+	if err != nil {
+		return err
+	}
+	return utils.WriteFile(filepath.Join(destinationDir, desc.Annotations[ocispec.AnnotationTitle]), b)
+}
+
+// PullPackageMetadata pulls the package metadata from the remote repository and saves it to `destinationDir`.
+func (o *OrasRemote) PullPackageMetadata(destinationDir string) error {
+	root, err := o.FetchRoot()
+	if err != nil {
+		return err
+	}
+	for _, path := range AlwaysPull {
+		desc := root.Locate(path)
+		if !o.isEmptyDescriptor(desc) {
+			err = o.PullFileLayer(desc, destinationDir)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	var pkg types.ZarfPackage
+	err = utils.ReadYaml(filepath.Join(destinationDir, config.ZarfYAML), &pkg)
+	if err != nil {
+		return err
+	}
+	if pkg.Metadata.AggregateChecksum != "" {
+		return utils.ValidatePackageChecksums(destinationDir, AlwaysPull)
+	}
```

Comment:
Noticed you wrap this with this empty string check in some places but don't in others.